Mathlib/LinearAlgebra/Matrix/FixedDetMatrices.lean

```diff
@@ -52,4 +53,228 @@ lemma smul_coe (m : R) (g : SpecialLinearGroup n R) (A : FixedDetMatrix n R m) :
+  decreasing_by
+    zify
+    exact reduce_aux h
+
+/--Map from `Δ m → Δ m` which reduces a FixedDetMatrix towards a representative element in reps. -/
+def reduce : Δ m → Δ m := fun A ↦
+  if |(A.1 1 0)| = 0 then
+    if  0 < A.1 0 0 then (T ^ (-(A.1 0 1/A.1 1 1))) • A else
+      (T ^ (-(-A.1 0 1/ -A.1 1 1))) • ( S • ( S • A)) --the -/- don't cancel with ℤ divs.
+  else
+    reduce (reduce_step A)
+  termination_by b => Int.natAbs (b.1 1 0)
+  decreasing_by
+      next a h =>
+      zify
+      exact reduce_aux h
+
+lemma reduce_of_pos {A : Δ m} (hc : |(A.1 1 0)| = 0) (ha : 0 < A.1 0 0) :
+    reduce A = (T ^ (-(A.1 0 1/A.1 1 1))) • A := by
+  rw [reduce]
+  simp only [zpow_neg, Int.ediv_neg, neg_neg] at *
+  simp_rw [if_pos hc, if_pos ha]
+
+lemma reduce_of_not_pos {A : Δ m} (hc : |(A.1 1 0)| = 0) (ha : ¬ 0 < A.1 0 0) :
+    reduce A = (T ^ (-(-A.1 0 1/ -A.1 1 1))) • ( S • ( S • A)) := by
+  rw [reduce]
+  simp only [abs_eq_zero, zpow_neg, Int.ediv_neg, neg_neg] at *
+  simp_rw [if_pos hc, if_neg ha]
+
+lemma reduce_of_reduce_step {A : Δ m} (hc : ¬ |(A.1 1 0)| = 0) :
+    reduce A = reduce (reduce_step A) := by
+  rw [reduce]
+  simp only [zpow_neg, Int.ediv_neg, neg_neg] at *
+  simp_rw [if_neg hc]
+
+private lemma A_c_eq_zero {A : Δ m} (ha : A.1 1 0 = 0) : A.1 0 0 * A.1 1 1 = m := by
+  simpa only [det_fin_two, ha, mul_zero, sub_zero] using A.2
+
+private lemma A_d_ne_zero {A : Δ m} (ha : |A.1 1 0| = 0) (hm : m ≠ 0) : A.1 1 1 ≠ 0 :=
+  right_ne_zero_of_mul (A_c_eq_zero (abs_eq_zero.mp ha) ▸ hm)
+
+private lemma A_a_ne_zero {A : Δ m} (ha : A.1 1 0 = 0) (hm : m ≠ 0) : A.1 0 0 ≠ 0 :=
+  left_ne_zero_of_mul (A_c_eq_zero ha ▸ hm)
+
+/--An auxiliary result bounding the size of the entries of the representatives in `reps`. -/
+lemma reps_entries_le_m' (hm : m ≠ 0) {A : Δ m} (h : A ∈ reps m) (i j : Fin 2) :
+    A.1 i j ∈ Finset.Icc (-|m|) |m|:= by
+  have h1 : 0 < |A.1 1 1| := abs_pos.mpr (A_d_ne_zero (abs_eq_zero.mpr h.left) hm)
+  have h2 : 0 < |A.1 0 0| := abs_pos.mpr (A_a_ne_zero h.left hm)
+  fin_cases i <;> fin_cases j
+  · simp only [← A_c_eq_zero h.1, Fin.zero_eta, Finset.mem_Icc, abs_mul]
+    constructor
+    · rw [neg_le]
+      exact (neg_le_abs (A.1 0 0)).trans ((le_mul_iff_one_le_right h2).mpr h1)
+    · exact (le_abs_self (A.1 0 0)).trans ((le_mul_iff_one_le_right h2).mpr h1)
+  · simp only [Fin.zero_eta, Fin.mk_one, Finset.mem_Icc]
+    constructor
+    · have := abs_pos.mpr hm
+      have := h.2.2.1
+      omega
+    · simp_rw [← A_c_eq_zero h.1, abs_mul]
+      exact (le_abs_self (A.1 0 1)).trans <| h.2.2.2.le.trans (le_mul_of_one_le_left h1.le h2)
+  · simp only [Fin.mk_one, Fin.zero_eta, h.1, Finset.mem_Icc, Left.neg_nonpos_iff, abs_nonneg,
+      and_self]
+  · simp only [Fin.mk_one, ← A_c_eq_zero h.1, abs_mul, Finset.mem_Icc]
+    constructor
+    · rw [neg_le]
+      exact (neg_le_abs (A.1 1 1)).trans ((le_mul_iff_one_le_left h1).mpr h2)
+    · exact (le_abs_self (A.1 1 1)).trans ((le_mul_iff_one_le_left h1).mpr h2)
```

Comment:
```suggestion
  suffices |A.1 i j| ≤ |m| from Finset.mem_Icc.mpr <| abs_le.mp this
  have h1 : 0 < |A.1 1 1| := abs_pos.mpr (A_d_ne_zero (abs_eq_zero.mpr h.left) hm)
  have h2 : 0 < |A.1 0 0| := abs_pos.mpr (A_a_ne_zero h.left hm)
  fin_cases i <;> fin_cases j
  · simpa only [Fin.zero_eta, ← A_c_eq_zero h.1, abs_mul]
    using (le_mul_iff_one_le_right h2).mpr h1
  · simpa only [Fin.zero_eta, Fin.mk_one, ← A_c_eq_zero h.1, abs_mul]
    using h.2.2.2.le.trans (le_mul_of_one_le_left h1.le h2)
  · simp only [Fin.mk_one, Fin.isValue, Fin.zero_eta, h.1, abs_zero, abs_nonneg]
  · simpa only [Fin.mk_one, ← A_c_eq_zero h.1, abs_mul] using (le_mul_iff_one_le_left h1).mpr h2
```
Going via the absolute value makes this easier, because there is some API available.